### PR TITLE
Requiring more type checking

### PR DESCRIPTION
**Pull request by liquidnya**
### Checklist

- [x] Have you followed the guidelines in our Contributing document?
- [x] Have you run `eslint` on the code and resolved any errors?
- [x] Have you run `npm test` and all tests are passing?
- [ ] Have you added new tests for any new functions created?
- [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you updated CHANGELOG.md to add your changed under the `[Unreleased]` heading?

### Description

Adds `plugin:@typescript-eslint/recommended-requiring-type-checking` to eslint.
Fixes around 100 issues that arised from these new rules.

Generally using zod to do type checking of unknown data (not just resulting from JSON) for you is very convenient to resolve some of those issues.

This PR also fixes that `dist.ts` and `build.ts` are now actually interpreted as TypeScript and not just JavaScript with TypeScript imports. This is done by introducing `tsconfig.swc.json` file that includes any files that are run by swc and then this file is configured using environment variables. Note that `src` is part of those files because `build.ts` is importing one shared source file. Also note that `tests` is not part of those files, because the test is setup with a different tsconfig file altogether.

This PR also introduces `npm run check` which will run `tsc` to check for compiler errors and this is configured using the `tsconfig.eslint.json` file now, such that tests, and all build scripts are also checked for compiler errors. This was necessary as eslint didn't find compiler errors in tests.

### Benefits

More type checks which can lead to less bugs.

### Potential drawbacks

New code will have to follow these rules. (more work)

### CODE REVIEW

**Review by demize — CHANGES_REQUESTED** [review]
As you mentioned in chat, we should probably also try to reduce the number of tsconfigs we have here. Doesn't hurt to take the time to do that now, and if we get it done now we won't have to worry about it later.

File: .github/workflows/node.js.yml [hunk 1]
```diff
@@ -27,7 +27,7 @@ jobs:
       - run: npm install --only=dev
       - run: ./node_modules/.bin/eslint src/
       - run: npx prettier --check .
-      - run: npx tsc --noemit
+      - run: npm run check
```
[the remark this answers is not in the document]
Comment (liquidnya) [reply .github/workflows/node.js.yml:29]: I moved `./node_modules/.bin/eslint src/` into `npm run check`, but I did not move prettier into this check, as I think it should be checked only for merging and/or committing, and should not be checked when distributing/building.
I also moved `npm run check` into `npm run dist-build` as well.
Or do you think the prettier should also be checked for distribution?

**Comment by liquidnya** [comment]
Now there is only one `tsconfig.json` file that includes all sources, tests, and build scripts.
In theory we could also create a `tsconfig.esbuild.json` file that could extend `tsconfig.json` and only include sources. That way if sources import from test or files outside `src` it might be stopped, but I did not think it was worth it.

File: .github/workflows/node.js.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (demize) [reply .github/workflows/node.js.yml:29]: Nope, I think this is good now.

**Review by demize — APPROVED** [review]
LGTM.

**Pull request merged**